clusterloader2/pkg/measurement/util/controlled_pods_indexer.go

```diff
@@ -40,8 +40,191 @@ type ControlledPodsIndexer struct {
+		podsSynced:            podsInformer.Informer().HasSynced,
+		rsUIDToState:          make(map[types.UID]*ReplicaSetState),
+		deploymentUIDToRsUIDs: make(map[types.UID]UIDSet),
+	}
+
+	podsInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+				rsUID, kind := cpi.getControllerInfo(obj)
+				if rsUID == "" || kind != "ReplicaSet" {
+					return
+				}
+				cpi.updatePodRefCountLocked(rsUID, 1)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				oldRsUID, oldKind := cpi.getControllerInfo(oldObj)
+				newRsUID, newKind := cpi.getControllerInfo(newObj)
+				if oldRsUID == newRsUID {
+					return
+				}
+				if oldKind == "ReplicaSet" {
+					cpi.updatePodRefCountLocked(oldRsUID, -1)
+				}
+				if newKind == "ReplicaSet" {
+					cpi.updatePodRefCountLocked(newRsUID, 1)
+				}
+				if oldKind == "ReplicaSet" {
+					cpi.clearRSDataIfPossibleLocked(oldRsUID)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				rsUID, kind := cpi.getControllerInfo(obj)
+				if rsUID == "" || kind != "ReplicaSet" {
+					return
+				}
+				cpi.updatePodRefCountLocked(rsUID, -1)
+				cpi.clearRSDataIfPossibleLocked(rsUID)
+			},
+		},
+	)
+	rsInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				deploymentUID, kind := cpi.getControllerInfo(obj)
+				if deploymentUID == "" || kind != "Deployment" {
+					return
+				}
+				cpi.handleIncomingRSAddEventLocked(obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				oldDeploymentUID, oldKind := cpi.getControllerInfo(oldObj)
+				newDeploymentUID, newKind := cpi.getControllerInfo(newObj)
+				if oldDeploymentUID == newDeploymentUID {
+					return
+				}
+				if oldKind == "Deployment" {
+					cpi.handleIncomingRSDeleteEventLocked(oldObj)
+				}
+				if newKind == "Deployment" {
+					cpi.handleIncomingRSAddEventLocked(newObj)
+				}
+				if oldKind == "Deployment" {
+					cpi.clearRSDataIfPossibleLocked(cpi.getObjUID(oldObj))
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				deploymentUID, kind := cpi.getControllerInfo(obj)
+				if deploymentUID == "" || kind != "Deployment" {
+					return
+				}
+				cpi.handleIncomingRSDeleteEventLocked(obj)
+				cpi.clearRSDataIfPossibleLocked(cpi.getObjUID(obj))
+			},
+		},
+	)
+
+	return cpi, nil
+}
+
+func (p *ControlledPodsIndexer) getControllerInfo(obj interface{}) (types.UID, string) {
+	metaAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		return "", ""
+	}
+	owner := metav1.GetControllerOf(metaAccessor)
+	if owner == nil {
+		return "", ""
+	}
+	return owner.UID, owner.Kind
```

Comment:
Nit: how about returning owner here?

IMHO it will increase visibility (i.e. nil owner means no owner, and it's clear what the first return value is and what the second value is + if you even need to access e.g. owner.Name, you won't need to change this function).